features/mbedtls/targets/TARGET_Silicon_Labs/crypto_sha.c

```diff
@@ -0,0 +1,449 @@
+    }
+
+    while( ilen >= 64 ) {
+        size_t blocks = ilen / 64;
+        crypto_sha_update_state( ctx->state, input, blocks, CRYPTO_SHA2 );
+        input += blocks * 64;
+        ilen  -= blocks * 64;
+    }
+
+    if( ilen > 0 ) {
+        memcpy( (void *) (ctx->buffer + left), input, ilen );
+    }
+}
+
+/*
+ * SHA-256 final digest
+ */
+void mbedtls_sha256_finish( mbedtls_sha256_context *ctx, unsigned char output[32] )
+{
+    uint32_t        last, padn;
+    uint32_t        high, low;
+    unsigned char   msglen[8];
+
+    high = ( ctx->total[0] >> 29 )
+         | ( ctx->total[1] <<  3 );
+    low  = ( ctx->total[0] <<  3 );
+
+    PUT_UINT32_BE( high, msglen, 0 );
+    PUT_UINT32_BE( low,  msglen, 4 );
+
+    last = ctx->total[0] & 0x3F;
+    padn = ( last < 56 ) ? ( 56 - last ) : ( 120 - last );
+
+    mbedtls_sha256_update( ctx, sha_padding, padn );
+    mbedtls_sha256_update( ctx, msglen, 8 );
+
+    for ( size_t i = 0; i < (ctx->is224 ? 28 : 32); i+=4) {
+        *((uint32_t*)(&output[i])) = __REV(ctx->state[i >> 2]);
+    }
+}
+#endif /* #if defined(MBEDTLS_SHA256_ALT) && defined(MBEDTLS_SHA256_C) */
+
+#if defined(MBEDTLS_SHA1_ALT) && defined(MBEDTLS_SHA1_C)
+
+/**
+ * \brief          Initialize SHA-1 context
+ */
+void mbedtls_sha1_init( mbedtls_sha1_context *ctx )
+{
+    if( ctx == NULL ) {
+        return;
+    }
+    memset( ctx, 0, sizeof( mbedtls_sha1_context ) );
+}
+
+/**
+ * \brief          Clear SHA-1 context
+ */
+void mbedtls_sha1_free( mbedtls_sha1_context *ctx )
+{
+    if( ctx == NULL ) {
+        return;
+    }
+    memset( ctx, 0, sizeof( mbedtls_sha1_context ) );
+}
+
+/**
+ * \brief          Clone (the state of) a SHA-1 context
+ */
+void mbedtls_sha1_clone( mbedtls_sha1_context *dst,
+                         const mbedtls_sha1_context *src )
+{
+    if ( dst != NULL && src != NULL ) {
+        *dst = *src;
+    }
+}
+
+/**
+ * \brief          SHA-1 context setup
+ *
+ * \param ctx      context to be initialized
+ */
+void mbedtls_sha1_starts( mbedtls_sha1_context *ctx )
+{
+    ctx->total[0] = 0;
+    ctx->total[1] = 0;
+    memcpy(ctx->state, init_state_sha1, 32);
+}
+
+/**
+ * \brief          SHA-1 process buffer
+ *
+ * \param ctx      SHA-1 context
+ * \param input    buffer holding the  data
+ * \param ilen     length of the input data
+ */
+void mbedtls_sha1_update( mbedtls_sha1_context *ctx,
+                          const unsigned char *input,
+                          size_t ilen )
+{
+    size_t fill;
+    uint32_t left;
+
+    if( ilen == 0 ) {
+        return;
+    }
+
+    left = ctx->total[0] & 0x3F;
+    fill = 64 - left;
+
+    ctx->total[0] += (uint32_t) ilen;
+    ctx->total[0] &= 0xFFFFFFFF;
+
+    if( ctx->total[0] < (uint32_t) ilen ) {
+        ctx->total[1]++;
+    }
+
+    if( left && ilen >= fill ) {
+        memcpy( (void *) (ctx->buffer + left), input, fill );
+        mbedtls_sha1_process( ctx, ctx->buffer );
+        input += fill;
+        ilen  -= fill;
+        left = 0;
```

Comment:
I think I understand what you did.
Isn't it better to have only one counter, that will only tell the size of the "residue buffer"? Once this "counter" will reach 64, you will call `crypto_sha_update_state` , and set this residue buffer size to zero